packages/shared/src/WithNativeNavigation.js

```diff
@@ -17,27 +19,41 @@ function withNativeNavigation<Props: {}>(
   moduleName: string,
 ): React.ComponentType<$Diff<Props, InjectedProps>> {
   return class WithNativeNavigation extends React.Component<*> {
-    currentIndex: number;
+    isEnabled: boolean;
+    lastCall: '' | 'disabled' | 'enabled';
 
     constructor(props) {
       super(props);
-      this.currentIndex = 0;
+      this.isEnabled = true;
+      this.lastCall = '';
     }
     onNavigationStateChange = (
       previousState: NavigationState,
       currentState: NavigationState,
     ) => {
-      if (currentState.index === 0) {
-        GestureController.enableGestures(moduleName);
-        this.currentIndex = 0;
-      } else if (currentState.index > 0) {
-        GestureController.disableGestures(moduleName);
-        this.currentIndex = currentState.index;
+      // Native gesture should only be enabled in the very first screen
+      // Also, there should not be modal (like first screen + modal)
+      const isDisabled =
+        currentState.routes.length > 1 ||
+        currentState.routes[0].routes.length > 1;
+
+      if (!isDisabled) {
+        if (this.lastCall !== 'enabled') {
```

Comment:
I don't like these nested ifs, can we get this in the same condition, or something like

```
if (disabled) callThisMEthod()
else callAnotherMethod()
```